kernel/src/process_loading.rs

```diff
@@ -451,17 +457,64 @@ pub trait ProcessLoadingAsync<'a> {
 
     /// Start the process loading operation.
     fn start(&self);
+
+    /// Check if the object of size `length` will be within flash bounds if placed
+    /// at address `offset`
+    fn check_if_within_flash_bounds(
+        &self,
+        offset: usize,
+        length: usize,
+    ) -> Result<(), ProcessBinaryError>;
+
+    /// Find available flash region for the new application.
+    fn check_flash_for_new_address(
+        &self,
+        new_app_size: usize,
+    ) -> Result<(usize, PaddingRequirement, usize, usize), ProcessBinaryError>;
+
+    /// Return the header slice for the new app.
+    fn check_new_binary_validity(
+        &self,
+        new_app_address: Option<usize>,
+    ) -> Result<(), ProcessBinaryError>;
+
+    /// Load applications from a new flash region. This automatically
+    /// starts the loading operation.
+    fn load_new_applications(
+        &self,
+        new_app_metadata: Option<(usize, usize)>,
+    ) -> Result<(), ProcessBinaryError>;
 }
 
 /// Operating mode of the loader.
 #[derive(Clone, Copy)]
 enum SequentialProcessLoaderMachineState {
     /// Phase of discovering `ProcessBinary` objects in flash.
     DiscoverProcessBinaries,
-    /// Phase of loading `ProcessBinary`s into `Process`s.
+    /// Phase of loading `ProcessBinary`s into `Process`es.
     LoadProcesses,
 }
 
+/// Metadata struct to hold list of process binary addresses.
+#[derive(Clone, Copy, Debug, Default)]
+struct ProcessBinaryMetadata {
+    index: usize,
+    process_binaries_start_addresses: [usize; MAX_PROCS],
+    process_binaries_end_addresses: [usize; MAX_PROCS],
+    new_app_address: usize,
+    new_app_size: usize,
```

Comment:
The `SequentialProcessLoaderMachine` shouldn't have a notion of a "new app". It understands how process binaries are stored in flash and how to load them, but it shouldn't track if something is new or not.